File: src/backend/internal/DummyMidiPort.cpp
#include "DummyMidiPort.h"
#include <algorithm>
#include "fmt/format.h"
#include <fmt/ranges.h>

#ifdef _WIN32
#undef min
#undef max
#endif

MidiStorageElem DummyMidiPort::get_event(uint32_t idx) const {
    try {
        if (!m_queued_msgs.empty()) {
            auto &m = m_queued_msgs.at(idx);
            ModuleLoggingEnabled<"Backend.DummyMidiPort">::log<log_level_debug>("Read queued midi message @ {}", m.time);
            return m;
        } else {
            auto &m = m_buffer_data.at(idx);
            ModuleLoggingEnabled<"Backend.DummyMidiPort">::log<log_level_debug>("Read buffer midi message @ {}", m.time);
            return m;
        }
    } catch (std::out_of_range &e) {
        throw std::runtime_error("Dummy midi port read error");
    }
}

void DummyMidiPort::write_event(MidiStorageElem event) {
    ModuleLoggingEnabled<"Backend.DummyMidiPort">::log<log_level_debug>("Write midi message value to internal buffer @ {}", event.time);
    m_buffer_data.push_back(event);
}

DummyMidiPort::DummyMidiPort(std::string name, shoop_port_direction_t direction, shoop_weak_ptr<DummyExternalConnections> external_connections)
    : MidiPort(true, true, true),
      DummyPort(name, direction, PortDataType::Midi, external_connections),
      WithCommandQueue(100, 1000, 1000) {}

unsigned DummyMidiPort::input_connectability() const {
    return (m_direction == ShoopPortDirection_Input) ? ShoopPortConnectability_External : ShoopPortConnectability_Internal;
}

unsigned DummyMidiPort::output_connectability() const {
    return (m_direction == ShoopPortDirection_Input) ? ShoopPortConnectability_Internal : ShoopPortConnectability_External;
}

void DummyMidiPort::clear_queues() {
    exec_process_thread_command([=]() {
        this->m_queued_msgs.clear();
        this->m_written_requested_msgs.clear();
        this->n_original_requested_frames = 0;
        this->n_requested_frames = 0;
    });
}

void DummyMidiPort::queue_msg(uint32_t size, uint32_t time, uint8_t const *data) {
    exec_process_thread_command([=]() {
        ModuleLoggingEnabled<"Backend.DummyMidiPort">::log<log_level_debug>("Queueing midi message @ {}", time);
        MidiStorageElem msg;
        msg.time = time;
        msg.size = size;
        memcpy(msg.bytes, data, size);
        m_queued_msgs.push_back(msg);
        std::stable_sort(this->m_queued_msgs.begin(), this->m_queued_msgs.end(), [](const MidiStorageElem &a, const MidiStorageElem &b) {
            return a.time < b.time;
        });
    });
}

bool DummyMidiPort::get_queue_empty() {
    bool is_empty = false;
    exec_process_thread_command([=, &is_empty]() {
        is_empty = this->m_queued_msgs.empty();
    });
    return is_empty;
}

void DummyMidiPort::request_data(uint32_t n_frames) {
    exec_process_thread_command([=]() {
        if (this->n_requested_frames > 0) {
            throw std::runtime_error("Previous request not yet completed");
        }
        ModuleLoggingEnabled<"Backend.DummyMidiPort">::log<log_level_debug_trace>("request {} frames", n_frames);
        this->n_requested_frames = n_frames;
        this->n_original_requested_frames = n_frames;
    });
}

void DummyMidiPort::PROC_prepare(uint32_t nframes) {
    PROC_handle_command_queue();
    m_buffer_data.clear();
    auto progress_by = n_processed_last_round.load();
    progress_by -= std::min(n_requested_frames.load(), progress_by);
    if (progress_by > 0 && m_queued_msgs.size() > 0) {
        this->ModuleLoggingEnabled<"Backend.DummyMidiPort">::log<log_level_debug_trace>("Progress queue by {}", progress_by);

        // The queue was used last pass and needs to be truncated now for the current pass.
        // (first erase msgs that will end up having a negative timestamp)
        std::erase_if(m_queued_msgs, [&, this](const MidiStorageElem &msg) {
            auto rval = msg.time < progress_by;
            if (rval) {
                this->ModuleLoggingEnabled<"Backend.DummyMidiPort">::log<log_level_debug_trace>("msg dropped from MIDI dummy input queue");
            }
            return rval;
        });
        std::for_each(m_queued_msgs.begin(), m_queued_msgs.end(), [&](MidiStorageElem &msg) {
            auto new_val = msg.time - progress_by;
            ModuleLoggingEnabled<"Backend.DummyMidiPort">::log<log_level_debug_trace>("msg in queue: time {} -> {}", msg.time, new_val);
            msg.time = new_val;
        });
    }
    n_processed_last_round = 0;
    current_buf_frames = nframes;
    MidiPort::PROC_prepare(nframes);
}

void DummyMidiPort::PROC_process(uint32_t nframes) {
    if (nframes > 0) {
        ModuleLoggingEnabled<"Backend.DummyMidiPort">::log<log_level_debug_trace>("Process {}", nframes);
    }
    if (m_direction == shoop_port_direction_t::ShoopPortDirection_Output) {
        std::stable_sort(m_buffer_data.begin(), m_buffer_data.end(), [](const MidiStorageElem &a, const MidiStorageElem &b) {
            return a.time < b.time;
        });
        if (!get_muted()) {
            for (auto &msg: m_buffer_data) {
                if (msg.time < n_requested_frames) {
                    uint32_t new_time = msg.time + (n_original_requested_frames - n_requested_frames);
                    ModuleLoggingEnabled<"Backend.DummyMidiPort">::log<log_level_debug>("Write midi message value to external queue @ {} -> {} {} {}", msg.time, new_time, n_original_requested_frames.load(), n_requested_frames.load());
                    MidiStorageElem out_msg = msg;
                    out_msg.time = new_time;
                    m_written_requested_msgs.push_back(out_msg);
                    // Track event through base
                    base().increment_output_events(1);
                }
            }
        } else {
            ModuleLoggingEnabled<"Backend.DummyMidiPort">::log<log_level_debug_trace>("{} messages not propagated to external due to being muted", m_buffer_data.size());
        }
    }
    n_processed_last_round = nframes;
    n_requested_frames -= std::min(nframes, n_requested_frames.load());
    MidiPort::PROC_process(nframes);
}

MidiReadableBuffer *
DummyMidiPort::PROC_get_read_output_data_buffer(uint32_t n_frames) {
    return this;
}

IMidiReadableBuffer *
DummyMidiPort::get_readable_buffer() {
    return this;
}

MidiWriteableBuffer *
DummyMidiPort::PROC_get_write_data_into_port_buffer(uint32_t n_frames) {
    current_buf_frames = n_frames;
    m_buffer_data.clear();
    return this;
}

IMidiWriteableBuffer *
DummyMidiPort::get_writeable_buffer() {
    return this;
}

uint32_t DummyMidiPort::n_events() const {
    uint32_t r = 0;
    if (!m_queued_msgs.empty()) {
        for (auto it = m_queued_msgs.begin(); it != m_queued_msgs.end(); ++it) {
            if (it->time < current_buf_frames) {
                r++;
            } else {
                break;
            }
        }
        return r;
    }
    return m_buffer_data.size();
}

std::vector<DummyMidiPort::StoredMessage> DummyMidiPort::get_written_requested_msgs() {
    std::vector<DummyMidiPort::StoredMessage> result;
    exec_process_thread_command([=, &result]() {
        ModuleLoggingEnabled<"Backend.DummyMidiPort">::log<log_level_debug>("Dequeue (have {} msgs)", this->m_written_requested_msgs.size());
        result = this->m_written_requested_msgs;
        this->m_written_requested_msgs.clear();
    });
    return result;
}

DummyMidiPort::~DummyMidiPort() { DummyPort::close(); }

File: src/rust/backend_rust/src/dummy_midi_port.rs
//! DummyMidiPort - A MIDI port implementation for testing and development in Rust.
//!
//! Inherits from MidiPort which uses composition (MidiPortBase) for state tracking
//! and ringbuffer functionality. Implements buffer interfaces for test use.
//!
//! Ported from C++ DummyMidiPort.h/cpp

use std::sync::atomic::{AtomicU32, Ordering};

use crate::midi_port::MidiPort;
use crate::midi_storage::MidiStorageElem;
use crate::midi_traits::{MidiReadableBuffer, MidiWritableBuffer};

/// Direction for the port
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PortDirection {
    Input,
    Output,
}

/// DummyMidiPort - A MIDI port implementation for testing and development.
///
/// Maintains internal buffers for test messages, supports queue/dequeue operations.
pub struct DummyMidiPort {
    // Base MIDI port (provides state tracking, ringbuffer, etc.)
    base: MidiPort,

    // Direction (input or output)
    direction: PortDirection,

    // Queued messages as external input to the port
    queued_msgs: Vec<MidiStorageElem>,

    // Buffer data for internal use
    buffer_data: Vec<MidiStorageElem>,

    // Messages written during the requested period
    pub written_requested_msgs: Vec<MidiStorageElem>,

    // Current buffer frame count
    current_buf_frames: AtomicU32,

    // Amount of frames requested for reading externally out of the port
    n_requested_frames: AtomicU32,

    // Number of frames processed in last round
    n_processed_last_round: AtomicU32,

    // Original requested frames
    n_original_requested_frames: AtomicU32,
}

impl DummyMidiPort {
    /// Create a new DummyMidiPort
    pub fn new(direction: PortDirection) -> Self {
        DummyMidiPort {
            base: MidiPort::new(true, true, true), // track_notes, track_controls, track_programs
            direction,
            queued_msgs: Vec::new(),
            buffer_data: Vec::new(),
            written_requested_msgs: Vec::new(),
            current_buf_frames: AtomicU32::new(0),
            n_requested_frames: AtomicU32::new(0),
            n_processed_last_round: AtomicU32::new(0),
            n_original_requested_frames: AtomicU32::new(0),
        }
    }

    /// Get reference to base MidiPort
    pub fn base(&self) -> &MidiPort {
        &self.base
    }

    /// Get mutable reference to base MidiPort
    pub fn base_mut(&mut self) -> &mut MidiPort {
        &mut self.base
    }

    /// Queue a message for input
    pub fn queue_msg(&mut self, time: u32, size: u16, data: &[u8]) {
        if let Some(elem) = MidiStorageElem::new(time, size, data) {
            self.queued_msgs.push(elem);
            // Sort by time
            self.queued_msgs.sort_by(|a, b| a.time.cmp(&b.time));
        }
    }

    /// Check if queue is empty
    pub fn get_queue_empty(&self) -> bool {
        self.queued_msgs.is_empty()
    }

    /// Clear all queues
    pub fn clear_queues(&mut self) {
        self.queued_msgs.clear();
        self.written_requested_msgs.clear();
        self.n_original_requested_frames.store(0, Ordering::SeqCst);
        self.n_requested_frames.store(0, Ordering::SeqCst);
    }

    /// Request a certain number of frames to be stored
    pub fn request_data(&mut self, n_frames: u32) {
        if self.n_requested_frames.load(Ordering::SeqCst) > 0 {
            panic!("Previous request not yet completed");
        }
        self.n_requested_frames.store(n_frames, Ordering::SeqCst);
        self.n_original_requested_frames
            .store(n_frames, Ordering::SeqCst);
    }

    /// Get messages written during the requested period
    pub fn get_written_requested_msgs(&mut self) -> Vec<MidiStorageElem> {
        let result = self.written_requested_msgs.clone();
        self.written_requested_msgs.clear();
        result
    }

    /// Get count of written messages
    pub fn get_n_written_requested_msgs(&self) -> u32 {
        self.written_requested_msgs.len() as u32
    }

    /// Get time of written message at index
    pub fn get_written_msg_time(&self, idx: u32) -> u32 {
        self.written_requested_msgs
            .get(idx as usize)
            .map(|e| e.time)
            .unwrap_or(u32::MAX)
    }

    /// Get size of written message at index
    pub fn get_written_msg_size(&self, idx: u32) -> u16 {
        self.written_requested_msgs
            .get(idx as usize)
            .map(|e| e.size)
            .unwrap_or(u16::MAX)
    }

    /// Get bytes of written message at index (fills output buffer)
    pub unsafe fn get_written_msg_bytes(&self, idx: u32, out: *mut u8, max_len: usize) {
        if let Some(elem) = self.written_requested_msgs.get(idx as usize) {
            let len = std::cmp::min(elem.size as usize, max_len);
            std::ptr::copy_nonoverlapping(elem.data().as_ptr(), out, len);
        }
    }

    /// Clear written messages
    pub fn clear_written_msgs(&mut self) {
        self.written_requested_msgs.clear();
    }

    /// Write an event directly (for FFI boundary)
    pub fn write_event_raw(&mut self, time: u32, size: u16, data: &[u8]) {
        if let Some(elem) = MidiStorageElem::new(time, size, data) {
            self.buffer_data.push(elem);
        }
    }

    /// Prepare for processing
    pub fn prepare(&mut self, nframes: u32) {
        self.buffer_data.clear();

        let progress_by = self.n_processed_last_round.load(Ordering::SeqCst);
        let mut new_progress = progress_by;
        new_progress = new_progress.saturating_sub(
            self.n_requested_frames
                .load(Ordering::SeqCst)
                .min(progress_by),
        );

        if new_progress > 0 && !self.queued_msgs.is_empty() {
            // Truncate old messages
            self.queued_msgs.retain(|msg| msg.time >= new_progress);
            // Adjust timestamps
            for msg in &mut self.queued_msgs {
                msg.time = msg.time.saturating_sub(new_progress);
            }
        }

        self.n_processed_last_round.store(0, Ordering::SeqCst);
        self.current_buf_frames.store(nframes, Ordering::SeqCst);
    }

    /// Process frames
    pub fn process(&mut self, nframes: u32) {
        if self.direction == PortDirection::Output {
            // Sort buffer data by time
            self.buffer_data.sort_by(|a, b| a.time.cmp(&b.time));

            if !self.base.get_muted() {
                let n_requested = self.n_requested_frames.load(Ordering::SeqCst);
                let n_original = self.n_original_requested_frames.load(Ordering::SeqCst);

                for msg in &self.buffer_data {
                    if msg.time < n_requested {
                        let new_time = msg.time + (n_original.saturating_sub(n_requested));
                        let mut out_msg = msg.clone();
                        out_msg.time = new_time;
                        self.written_requested_msgs.push(out_msg);
                        // Track event through base
                        self.base.increment_output_events(1);
                    }
                }
            }
        }

        self.n_processed_last_round.store(nframes, Ordering::SeqCst);
        let n_req = self.n_requested_frames.load(Ordering::SeqCst);
        self.n_requested_frames
            .store(n_req.saturating_sub(nframes.min(n_req)), Ordering::SeqCst);
    }

    /// Get the direction
    pub fn direction(&self) -> PortDirection {
        self.direction
    }
}

// Implement MidiReadableBuffer for DummyMidiPort
impl MidiReadableBuffer for DummyMidiPort {
    fn n_events(&self) -> u32 {
        if !self.queued_msgs.is_empty() {
            let mut count = 0u32;
            for msg in &self.queued_msgs {
                if msg.time < self.current_buf_frames.load(Ordering::SeqCst) {
                    count += 1;
                } else {
                    break;
                }
            }
            return count;
        }
        self.buffer_data.len() as u32
    }

    fn get_event(&self, idx: u32) -> Option<MidiStorageElem> {
        if !self.queued_msgs.is_empty() {
            return self.queued_msgs.get(idx as usize).cloned();
        }
        self.buffer_data.get(idx as usize).cloned()
    }
}

// Implement MidiWritableBuffer for DummyMidiPort
impl MidiWritableBuffer for DummyMidiPort {
    fn write_event(&mut self, event: MidiStorageElem) -> bool {
        self.buffer_data.push(event);
        true
    }
}

// Delegation methods for CXX bridge compatibility
impl DummyMidiPort {
    /// Get mute state
    pub fn get_muted(&self) -> bool {
        self.base.get_muted()
    }

    /// Set mute state
    pub fn set_muted(&mut self, muted: bool) {
        self.base.set_muted(muted);
    }

    /// Get number of active notes
    pub fn n_notes_active(&self) -> u32 {
        self.base.n_notes_active()
    }

    /// Get number of events (from MidiReadableBuffer trait)
    pub fn n_events(&self) -> u32 {
        MidiReadableBuffer::n_events(self)
    }

    /// Get event time at index (from MidiReadableBuffer trait)
    pub fn get_event_time(&self, idx: u32) -> u32 {
        MidiReadableBuffer::get_event(self, idx)
            .map(|e| e.time)
            .unwrap_or(u32::MAX)
    }

    /// Get event size at index (from MidiReadableBuffer trait)
    pub fn get_event_size(&self, idx: u32) -> u16 {
        MidiReadableBuffer::get_event(self, idx)
            .map(|e| e.size)
            .unwrap_or(u16::MAX)
    }

    /// Get event bytes at index (from MidiReadableBuffer trait)
    pub unsafe fn get_event_bytes(&self, idx: u32, out: *mut u8, max_len: usize) {
        if let Some(elem) = MidiReadableBuffer::get_event(self, idx) {
            let len = std::cmp::min(elem.size as usize, max_len);
            std::ptr::copy_nonoverlapping(elem.data().as_ptr(), out, len);
        }
    }

    /// Write event (for CXX bridge - delegates to MidiWritableBuffer trait)
    pub fn write_event_cxx(&mut self, time: u32, size: u16, data: &[u8]) {
        self.write_event_raw(time, size, data);
    }

    /// Dummy write event for CXX bridge (same as write_event_cxx)
    pub fn dummy_write_event(&mut self, time: u32, size: u16, data: &[u8]) {
        self.write_event_cxx(time, size, data);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_dummy_midi_port_creation() {
        let port = DummyMidiPort::new(PortDirection::Input);
        assert!(!port.base.get_muted());
    }

    #[test]
    fn test_queue_message() {
        let mut port = DummyMidiPort::new(PortDirection::Input);
        port.queue_msg(100, 3, &[0x90, 0x3C, 0x7F]);
        assert!(!port.get_queue_empty());
    }

    #[test]
    fn test_clear_queues() {
        let mut port = DummyMidiPort::new(PortDirection::Input);
        port.queue_msg(100, 3, &[0x90, 0x3C, 0x7F]);
        assert!(!port.get_queue_empty());
        port.clear_queues();
        assert!(port.get_queue_empty());
    }

    #[test]
    fn test_readable_buffer() {
        let mut port = DummyMidiPort::new(PortDirection::Input);
        port.queue_msg(50, 3, &[0x90, 0x3C, 0x7F]);
        port.queue_msg(100, 3, &[0x90, 0x40, 0x7F]);

        port.prepare(200);
        assert_eq!(port.n_events(), 2);
    }

    #[test]
    fn test_writable_buffer() {
        let mut port = DummyMidiPort::new(PortDirection::Output);
        let event = MidiStorageElem::new(50, 3, &[0x90, 0x3C, 0x7F]).unwrap();
        assert!(port.write_event(event));
        assert_eq!(port.n_events(), 1);
    }

    #[test]
    fn test_request_data() {
        let mut port = DummyMidiPort::new(PortDirection::Input);
        port.request_data(1024);
        assert_eq!(port.n_requested_frames.load(Ordering::SeqCst), 1024);
    }
}
